src/account/EphemeralKeyPair.ts

```diff
@@ -0,0 +1,158 @@
+// SPDX-License-Identifier: Apache-2.0
+
+import { randomBytes } from "@noble/hashes/utils";
+
+import { Ed25519PrivateKey, EphemeralPublicKey, EphemeralSignature, PrivateKey } from "../core/crypto";
+import { Hex } from "../core/hex";
+import { bytesToBigIntLE, padAndPackBytesWithLen, poseidonHash } from "../core/crypto/poseidon";
+import { EphemeralPublicKeyVariant, HexInput } from "../types";
+import { Deserializer, Serializable, Serializer } from "../bcs";
+import { floorToWholeHour, nowInSeconds } from "../utils/helpers";
+
+const TWO_WEEKS_IN_SECONDS = 1_209_600;
+
+/**
+ * A class which contains a key pair that is used in signing transactions via the Keyless authentication scheme. This key pair
+ * is ephemeral and has an expiration time.  For more details on how this class is used -
+ * https://aptos.dev/guides/keyless-accounts/#1-present-the-user-with-a-sign-in-with-idp-button-on-the-ui
+ */
+export class EphemeralKeyPair extends Serializable {
+  static readonly BLINDER_LENGTH: number = 31;
+
+  /**
+   * A byte array of length BLINDER_LENGTH used to obfuscate the public key from the IdP.
+   * Used in calculating the nonce passed to the IdP and as a secret witness in proof generation.
+   */
+  readonly blinder: Uint8Array;
+
+  /**
+   * A timestamp in seconds indicating when the ephemeral key pair is expired.  After expiry, a new
+   * EphemeralKeyPair must be generated and a new JWT needs to be created.
+   */
+  readonly expiryDateSecs: number;
+
+  /**
+   * The value passed to the IdP when the user authenticates.  It comprises of a hash of the
+   * ephermeral public key, expiry date, and blinder.
+   */
+  readonly nonce: string;
+
+  /**
+   * A private key used to sign transactions.  This private key is not tied to any account on the chain as it
+   * is ephemeral (not permanent) in nature.
+   */
+  private privateKey: PrivateKey;
+
+  /**
+   * A public key used to verify transactions.  This public key is not tied to any account on the chain as it
+   * is ephemeral (not permanent) in nature.
+   */
+  private publicKey: EphemeralPublicKey;
+
+  constructor(args: { privateKey: PrivateKey; expiryDateSecs?: number; blinder?: HexInput }) {
+    super();
+    const { privateKey, expiryDateSecs, blinder } = args;
+    this.privateKey = privateKey;
+    this.publicKey = new EphemeralPublicKey(privateKey.publicKey());
+    // By default, we set the expiry date to be two weeks in the future floored to the nearest hour
+    this.expiryDateSecs = expiryDateSecs || floorToWholeHour(nowInSeconds() + TWO_WEEKS_IN_SECONDS);
+    // Generate the blinder if not provided
+    this.blinder = blinder !== undefined ? Hex.fromHexInput(blinder).toUint8Array() : generateBlinder();
+    // Calculate the nonce
+    this.nonce = this.generateNonce();
+  }
+
+  /**
+   * Returns the public key of the key pair.
+   * @return EphemeralPublicKey
+   */
+  getPublicKey(): EphemeralPublicKey {
+    return this.publicKey;
+  }
```

Comment:
So @alinush was concerned about the field's misuse and that somebody might think it is the address of an account.

This provides a layer of indirection that makes it harder of confuse (since else where you can access the public key attribute directly), but I also prefer letting it be readonly...

Reply:
Just to confirm: devs will not be able to feed an `EphemeralPublicKey` into a preeexisting call and obtain an address, right?

Reply:
(I assume we have a type-safe way to derive addresses in the SDK for any of our supported signature schemes?)

Reply:
Yeah I don't think it's needed, `PublicKey` and `AccountAddress` are two very different types that are not compatible with each other.

Reply:
As long as there's no easy way to convert `EphemeralPublicKey` to `AccountAddress` in the SDK, we have a much needed layer of protection.

(@hardsetting, it's `EphemeralPublicKey` not `PublicKey`, right?)